src/libmongoc/src/mongoc/mongoc-structured-log.c

```diff
@@ -0,0 +1,772 @@
+{
+   const char *path = getenv ("MONGODB_LOG_PATH");
+   if (!path || !strcmp (path, "stderr")) {
+      return stderr;
+   }
+   if (!strcmp (path, "stdout")) {
+      return stdout;
+   }
+   FILE *file = fopen (path, "a");
+   if (!file) {
+      MONGOC_WARNING ("Cannot open log file '%s' for writing. Logging to stderr instead.", path);
+      return stderr;
+   }
+   return file;
+}
+
+static FILE *
+_mongoc_structured_log_get_stream (void)
+{
+   // Not re-entrant; protected by the default_handler.stream_mutex.
+   FILE *log_stream = gStructuredLog.default_handler.stream;
+   if (log_stream) {
+      return log_stream;
+   }
+   // Note that log_stream may be the global stderr/stdout streams,
+   // or an allocated FILE that is never closed.
+   log_stream = _mongoc_structured_log_open_stream ();
+   gStructuredLog.default_handler.stream = log_stream;
+   return log_stream;
+}
+
+static void
+mongoc_structured_log_default_handler (const mongoc_structured_log_entry_t *entry, void *user_data)
+{
+   // We can serialize the message before taking the stream lock
+   bson_t *bson_message = mongoc_structured_log_entry_message_as_bson (entry);
+   char *json_message = bson_as_relaxed_extended_json (bson_message, NULL);
+   bson_destroy (bson_message);
+
+   const char *level_name = mongoc_structured_log_get_level_name (mongoc_structured_log_entry_get_level (entry));
+   const char *component_name =
+      mongoc_structured_log_get_component_name (mongoc_structured_log_entry_get_component (entry));
+
+   bson_mutex_lock (&gStructuredLog.default_handler.stream_mutex);
+   fprintf (_mongoc_structured_log_get_stream (), "MONGODB_LOG %s %s %s\n", level_name, component_name, json_message);
+   bson_mutex_unlock (&gStructuredLog.default_handler.stream_mutex);
+
+   bson_free (json_message);
+}
+
+static BSON_ONCE_FUN (_mongoc_structured_log_init_once)
+{
+   bson_shared_mutex_init (&gStructuredLog.handler.mutex);
+   bson_mutex_init (&gStructuredLog.default_handler.stream_mutex);
+   bson_mutex_init (&gStructuredLog.combined_table.build_mutex);
+
+   gStructuredLog.max_document_length = _mongoc_structured_log_get_max_document_length_from_env ();
+
+   bson_mutex_lock (&gStructuredLog.combined_table.build_mutex);
+
+   _mongoc_structured_log_set_max_level_for_all_components (MONGOC_STRUCTURED_LOG_DEFAULT_LEVEL);
+   _mongoc_structured_log_set_max_levels_from_env ();
+
+   // The default handler replaces MONGOC_STRUCTURED_LOG_STATE_MAYBE_UNINITIALIZED.
+   // Other threads may immediately submit log entries, so this must happen after
+   // log level setup and just before we unlock the build_mutex and return.
+   _mongoc_structured_log_set_handler (mongoc_structured_log_default_handler, NULL);
+   _mongoc_structured_log_update_level_plus_state ();
+
+   bson_mutex_unlock (&gStructuredLog.combined_table.build_mutex);
+
+   BSON_ONCE_RETURN;
+}
+
+char *
+mongoc_structured_log_document_to_json (const bson_t *document, size_t *length)
```

Comment:
It's not especially necessary no, we could remove it. It's an internal API from Andreas' version of the PR. I thought it was somewhat useful, but possibly it's more confounding than useful.

Reply:
No objections if it's private. I was mainly asking if it should be a static function within this file instead of appearing in the private header. I'll defer to your judgement.

Reply:
In the non-global redesign I ended up reducing the visibility of this method and also changing its name to emphasize that it's only for use with inner documents within a log message. (_mongoc_structured_log_inner_document_to_json)

It now has a clear dependency on the logging options, which should make it hard to accidentally misuse even if we want to make it more visible later.